### ISSUE TITLE
Having trouble with `Quectel RM520N-GL` in Thinkpad X1 Carbon 12th gen (Using Linux `6.9.2`)

### ISSUE THREAD

**Issue by ndom91** [lenovo/lenovo-wwan-unlock#17]
So first of all, I'm a little confused. I have a German 12th gen X1 Carbon with the RM520N Quectel 5G modem - do I even need this unlock procedure? It seems to me that it's only needed for American market targeted devices that might have been locked due to FCC rules?

Anyway, that being said, I'm having a hell of a time getting my system to recognize this modem..

I'm running NixOS with a 6.9.2 Kernel. I see in the kernel mailing lists that support for this is supposed to have been shipped in ~6.6 kernel. I also have ModemManager v1.22, which should be recent enough according to this README.

I've tried enabling all sorts of kernel modules (i.e. `boot.initrd.kernelModules = [ "xe" "mhi_pci_generic" "wwan" "wwan_hwsim" "usbserial" "qcserial" "qmi_wwan" "option" "cdc_mbim" "mhi_net" "quatech2" ];`) and all I've gotten is these devices to show up under `/dev/*`: `wwan0at0`, `wwan0at1`, `wwan1at0`, `wwan1at1`. Unfortnuately there is no additional `ttyUSB*` device, or `qmi*` or `cdc-*` device. So the modem manager CLI doens't reconize any modem. Also `lsusb` and `lspci` don't list anything related to Quectel or WWAN or anything of the sort.

What else could I be missing? :thinking:

**Comment by ndom91** [comment]
Okay so turns out I got the SKU that ships with a Nano SIM slot, but no modem installed :facepalm:. After fiddling around with kernel modules for 2 days, I decided to open it up and confirm the modem part # and what not and saw the slot was empty hahah.

But who sells this kind of SKU with a SIM Slot on a laptop, "5G" in the title, and then doesn't actually include a modem :shrug:

**Comment by lxwulf** [comment]
**Free tip:**

You could test what is built-in with `lspci`. For your or our specific situation, you could run the following command:

```bash
lspci | grep Wireless\ \controller
```
This gives you back what is inside your laptop. When the output is nothing, then nothing is built-in.
At least in the meaning of a `wireless controller`.

On my side, it looks like so:

```bash
ok@thinkpadx395:~$ lspci | grep Wireless\ \controller
04:00.0 Wireless controller [0d40]: Intel Corporation XMM7360 LTE Advanced Modem (rev 01)
```

This is, as you may have recognized, a `Thinkpad x395`.

> Of course you could also set the `search content` in " ", like so:
>
> ```bash
> lspci | grep "Wireless controller"
> ```

> **Info:**
> Your situation with no module built-in could be a country specific problem, because different countries has different regulations to frequencies and radio signals. So many companies implement the function to have an WWAN module but do not build one in because else they need to fiddle around with the law and regulations of the specific country. It's as always a very complicated manner, and of course money plays also a reason in this.

**Comment by ndom91** [comment]
Yeah as I mentioned in the first post, `lspci` and `lsusb` were both not showing anything, so I assumed it was a kernel module issue.

**Comment by lxwulf** [comment]
Oh, sorry, I missed that! 🙈

# TL;DR!
But as you already said, nonetheless if any driver is loaded or not, it will show the attached device in the PCIe slot.

---

# More Info

My WWAN module has no driver, but still is recognized by the system with the hardware name as mentioned above.

Keep in mind that sometimes the hardware has a different naming scheme. Like my example, the XMM7360 is mostly sold with the name Fibocom L850-GL. Despite this curiosity, this is also not always the case. But in the WWAN module section, again it mostly the case.

The most time, there is a hardware supplier and one buyer which mostly then modify the basic structure to adapt it to his system which then gets sold to the customer. This could be me and you, but mostly it is other companies like Lenovo.

This is just a brief summary. Just wanted to give a better overview or little insight. 😊
